src/theory/quantifiers/sygus/term_database_sygus.cpp

```diff
@@ -725,6 +726,28 @@ SygusTypeInfo& TermDbSygus::getTypeInfo(TypeNode tn)
   return d_tinfo[tn];
 }
 
+Node TermDbSygus::rewriteNode(Node n) const
+{
+  Node res = Rewriter::rewrite(n);
+  if (options::sygusRecFun())
+  {
+    if (d_funDefEval->hasDefinitions())
+    {
+      // If recursive functions are enabled, then we use the recursive function
+      // evaluation utility.
+      Node fres = d_funDefEval->evaluate(res);
+      if (!fres.isNull())
+      {
+        return fres;
+      }
+      // It may have failed, in which case there are undefined symbols in res.
+      // In this case, we revert to the result of rewriting in the return
+      // statement below.
```

Comment:
I'm a bit confused by this. Why do you discard the result of performing evaluation if the whole term is not reduced to a constant? Wouldn't you have simplified at least part of it via the evaluations?

Reply:
It's currently all or nothing. It isn't totally clear if partially evaluating has any use here: we are only using this in the context of evaluation. A partially evaluated node may actually be worse (think of a case that a function expands/terminates with some huge expression that we don't want to keep around).

Reply:
Yeah... I wonder. I think that the evaluator should be akin to the rewriter, in the sense that we always use it and in general it helps. If it does not it's an issue of the user. It might difficult for us to determine a priori whether things well help or not, so we assume they will (like we do with quantifiers).